src/stc/PlatWX.cpp

```diff
@@ -3026,24 +3075,85 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
     int topGap = m_textTopGap;
     int leftGap = TextBoxFromClientEdge() + m_textBoxToTextGap;
 
-    wxDCTextColourChanger tcc(dc);
+    wxColour textCol;
 
     if ( IsSelected(n) )
-        tcc.Set(m_visualData->GetHighlightTextColour());
+        textCol = m_visualData->GetHighlightTextColour();
     else if ( static_cast<int>(n) == m_currentRow )
-        tcc.Set(m_visualData->GetCurrentTextColour());
+        textCol = m_visualData->GetCurrentTextColour();
+    else
+        textCol = m_visualData->GetTextColour();
+
+    if ( m_technology == wxSTC_TECHNOLOGY_DIRECTWRITE )
+    {
+    #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+        SurfaceD2D surface;
+        surface.Init(&dc, GetGrandParent());
+
+        wxString ellipsizedLabel = label;
+
+        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
+        SurfaceFontDataD2D* newSurfaceData =
+            new SurfaceFontDataD2D(*m_surfaceFontData);
+        fontCopy->SetSurfaceFontData(newSurfaceData);
+        Font tempFont;
+        tempFont.SetID(fontCopy);
+
+        wxCharBuffer buffer = wx2stc(label);
+        int curWidth = surface.WidthText(tempFont, buffer.data(),
+                                          wx2stclen(wxString(),buffer));
+
+        int maxWidth = rect.GetWidth() - leftGap;
+
+        if ( curWidth > maxWidth )
```

Comment:
This is also pretty minor, but it's a bit confusing to see comparisons in both directions. Could we write this as a loop to avoid this? I.e. something like
```cpp
for ( int i = label.length(); curWidth > maxWidth && i; --i )
{
    ...
}
```
?

Reply:
I made both of these changes below.